File: Schema/SchemaDeploy/tables/Group/CGroupList.regenerated.cs
using System;
using System.Data;
using System.Collections;
using System.Collections.Generic;

using Framework;

namespace SchemaDeploy
{
    [Serializable()]
    public partial class CGroupList : List<CGroup>
    {
        #region Constructors
        //Basic constructor
        public CGroupList() : base() {}
       
        //More efficient memory-allocation if size is known
        public CGroupList(int capacity) : base(capacity) {}
        
        //Shares the index (if its already been computed)
        public CGroupList(CGroupList list) : base(list.Count)
        {
            foreach (CGroup i in list)
                base.Add(i);
        }

        //Generic list (eg. from paging control), have to assume type
        public CGroupList(IList list) : base(list.Count)
        {
            foreach (CGroup i in list)
                base.Add(i);
        }
        #endregion
        
        #region Top/Bottom/Page
        public CGroupList Top(int count)
        {
            if (count >= this.Count)
                return this;
            return Page(count, 0); 
        }
        public CGroupList Bottom(int count)   
        {
            if (count > this.Count)
                count = this.Count;
            return new CGroupList(this.GetRange(this.Count - count - 1, count));
        }
        public CGroupList Page(int pageSize, int pageIndex)
        {
            return new CGroupList( CUtilities.Page(this, pageSize, pageIndex) );
        }
        #endregion

        #region BulkEditLogic
        public bool HaveSameValue(string propertyName)               { return CReflection.HaveSameValue(this, propertyName); }
        public void SetSameValue( string propertyName, object value) {        CReflection.SetSameValue( this, propertyName, value); }
        #endregion

        #region SortBy
        //Public
        public CGroupList SortBy(string propertyName) { return SortBy(propertyName, false); }
        public CGroupList SortBy(string propertyName, bool descending)
        {
            CGroupList copy = new CGroupList(this);
            if (this.Count == 0)    return copy;
            copy.Sort(new CGroupList_SortBy(propertyName, descending, this));
            return copy;
        }
        //Private 
        private class CGroupList_SortBy : CReflection.GenericSortBy, IComparer<CGroup>
        {
            public CGroupList_SortBy(string propertyName, bool descending, IList list) : base(propertyName, descending, list) { }
            public int Compare(CGroup x, CGroup y) { return base.Compare(x, y); }
        }
        #endregion

        #region SaveAll/DeleteAll
        //Use default connection (may be overridden in base class)
        public void SaveAll()   {   if (this.Count > 0) { SaveAll(  this[0].DataSrc);    }    }
        public void DeleteAll() {   if (this.Count > 0) { DeleteAll(this[0].DataSrc);    }    }

        //Use connection supplied
        public void SaveAll(  CDataSrc dataSrc) {   dataSrc.BulkSave(this);    }
        public void DeleteAll(CDataSrc dataSrc) {   dataSrc.BulkDelete(this);  }

        //Use transaction supplied
        public void SaveAll(  IDbTransaction txOrNull)    {   foreach (CGroup i in this) {   i.Save(  txOrNull);   }   }
        public void DeleteAll(IDbTransaction txOrNull)    {   foreach (CGroup i in this) {   i.Delete(txOrNull);   }   }

        //Use a specified isolation level
        public void SaveAll(  IsolationLevel txIsolationLevel)   {   if (this.Count > 0) { SaveAll(  this[0].DataSrc, txIsolationLevel);  }    }
        public void DeleteAll(IsolationLevel txIsolationLevel)   {   if (this.Count > 0) { DeleteAll(this[0].DataSrc, txIsolationLevel);  }    }

        //Use a specified connection and isolation level
        public void SaveAll(  CDataSrc dataSrc, IsolationLevel txIsolationLevel)   {   dataSrc.BulkSave(  this, txIsolationLevel);  }
        public void DeleteAll(CDataSrc dataSrc, IsolationLevel txIsolationLevel)   {   dataSrc.BulkDelete(this, txIsolationLevel);  }
        #endregion

        #region List of Ids
        List<int> _ids;
        public List<int> Ids
        {
            get
            {
                if (null == _ids)
                {
                    List<int> ids = new List<int>(this.Count);
                    foreach (CGroup i in this)
                        ids.Add(i.GroupId);
                    _ids = ids;
                }
                return _ids;
            }
        }
        public CGroupList GetByIds(List<int> ids)
        {
            CGroupList list = new CGroupList(ids.Count);
            foreach (int id in ids)
                if (null != GetById(id))
                    list.Add(GetById(id));
            return list;
        }
        #endregion
        
        #region Cache-Control
        //Main Logic
        public new void Add(CGroup item)
        {
            if (null != _index && ! _index.ContainsKey(item.GroupId))
                _index[item.GroupId] = item;
            base.Add(item);
        }
        public new void Remove(CGroup item)
        {
            if (null != _index && _index.ContainsKey(item.GroupId))
                _index.Remove(item.GroupId);
            base.Remove(item);
        }
    
        //Supplementary List Overloads
        public void Add(   IList<CGroup> itemsToAdd)    {   foreach (CGroup i in itemsToAdd)   { Add(   i); }   }
        public void Remove(IList<CGroup> itemsToRemove) {   foreach (CGroup i in itemsToRemove){ Remove(i); }   }
        #endregion
        
        #region Main Index (on GroupId)
        public CGroup GetById(int groupId)
        {
            CGroup c = null;
            Index.TryGetValue(groupId, out c);
            return c;
        }
        [NonSerialized]
        private Dictionary<int,CGroup> _index;
        private Dictionary<int,CGroup> Index
        {
            get
            {
                if (null != _index)
                    if (_index.Count == this.Count)
                        return _index;

                    _index = new Dictionary<int,CGroup>(this.Count);
                        foreach (CGroup i in this) 
                            _index[i.GroupId] = i;
                return _index;
            }
        }
        #endregion
            
        #region Foreign-Key Indices (Subsets)
        //Index by GroupAppId
        public CGroupList GetByAppId(int appId)
        {
            CGroupList temp = null;
            if (! IndexByAppId.TryGetValue(appId, out temp))
            {
                temp = new CGroupList();
                IndexByAppId[appId] = temp;
            }
            return temp;
        }

        [NonSerialized]
        private Dictionary<int, CGroupList> _indexByAppId;
        private Dictionary<int, CGroupList> IndexByAppId
        {
            get
            {
                if (null == _indexByAppId)
                {
                    Dictionary<int, CGroupList> index = new Dictionary<int, CGroupList>();
                    CGroupList temp = null;
                    foreach (CGroup i in this)
                    {
                        if (! index.TryGetValue(i.GroupAppId, out temp))
                        {
                            temp = new CGroupList();
                            index[i.GroupAppId] = temp;
                        }
                        temp.Add(i);
                    }
                    _indexByAppId = index;
                }
                return _indexByAppId;
            }
        }
        #endregion

    }
}

File: Schema/SchemaConfig/tables/Group/CGroupList.regenerated.vb
Imports System
Imports System.Data
Imports System.Collections.Generic

Imports Framework

<Serializable(), CLSCompliant(True)> _
Partial Public Class CGroupList : Inherits List(Of CGroup)

#Region "Constructors"
    'Basic constructor
    Public Sub New()
        MyBase.New()  
    End Sub
    
    'More efficient memory-allocation if size is known
    Public Sub New(ByVal capacity As Integer)
        MyBase.New(capacity)  
    End Sub
    
    'Shares the index (if its already been computed)
    Public Sub New(ByVal list As CGroupList)
        MyBase.New(list)
        _index = list._index 
    End Sub

    'Generic list (eg. from paging control), have to assume type
    Public Sub New(ByVal list As IList)
        MyBase.New(list.Count)
        For Each i As CGroup In list
            MyBase.Add(i)
        Next
    End Sub
#End Region

#Region "Top/Bottom"
    Public Function Top(ByVal count As Integer) As CGroupList
        Return New CGroupList(CUtilities.Page(Me, count, 0))
    End Function

    Public Function Bottom(ByVal count As Integer) As CGroupList
        If count > Me.Count Then count = Me.Count
        Return New CGroupList(Me.GetRange(Me.Count - count - 1, count))
    End Function
#End Region

#Region "SaveAll/DeleteAll"
    'Use default connection (may be overridden in base class)
    Public Sub SaveAll()
        If Me.Count = 0 Then Exit Sub
        SaveAll(Me(0).DataSrc)
    End Sub
    Public Sub DeleteAll()
        If Me.Count = 0 Then Exit Sub
        DeleteAll(Me(0).DataSrc)
    End Sub
    
    'Use connection supplied
    Public Sub SaveAll(dataSrc as CDatasrc)
        dataSrc.BulkSave(Me)
    End Sub
    Public Sub DeleteAll(dataSrc as CDatasrc)
        dataSrc.BulkDelete(Me)
    End Sub

    'Use transaction supplied
    Public Sub SaveAll(ByVal txOrNull As IDbTransaction)
        For Each i As CGroup In Me
            i.Save(txOrNull)
        Next
    End Sub
    Public Sub DeleteAll(ByVal txOrNull As IDbTransaction)
        For Each i As CGroup In Me
            i.Delete(txOrNull)
        Next
    End Sub

    'Use a specified isolation level
    Public Sub SaveAll(ByVal txIsolationLevel As IsolationLevel)
        If Me.Count = 0 Then Exit Sub
        SaveAll(Me(0).DataSrc, txIsolationLevel)
    End Sub
    Public Sub DeleteAll(ByVal txIsolationLevel As IsolationLevel)
        If Me.Count = 0 Then Exit Sub
        DeleteAll(Me(0).DataSrc, txIsolationLevel)
    End Sub
    
    'Use a specified connection and isolation level
    Public Sub SaveAll(ByVal dataSrc as CDatasrc, ByVal txIsolationLevel As IsolationLevel)
        dataSrc.BulkSave(Me, txIsolationLevel)
    End Sub
    Public Sub DeleteAll(ByVal dataSrc as CDatasrc, ByVal txIsolationLevel As IsolationLevel)
        dataSrc.BulkDelete(Me, txIsolationLevel)
    End Sub
#End Region

#Region "Cache-Control"
    Public Shadows Sub Add(ByVal item As CGroup)
        If Not IsNothing(_index) Then
            If Not _index.ContainsKey(item.GroupId) Then
                _index(item.GroupId) = item
            End If
        End If
        MyBase.Add(item)
    End Sub
    Public Shadows Sub Remove(ByVal item As CGroup)
        If Not IsNothing(_index) Then
            If _index.ContainsKey(item.GroupId) Then
                _index.Remove(item.GroupId)
            End If
        End If
        MyBase.Remove(item)
    End Sub
#End Region

#Region "Main Index (on GroupId)"
    'Only uncomment if type is NOT an Integer (e.g. string/guid)
    'Default Public Overloads ReadOnly Property Item(ByVal groupId As Integer) As CGroup
    '    Get
    '        Return GetById(groupId)
    '    End Get
    'End Property
    Public Function GetById(ByVal groupId As Integer) As CGroup
        Dim c As CGroup = Nothing
        Index.TryGetValue(groupId, c)
        Return c
    End Function
    <NonSerialized()> _
    Private _index As Dictionary(Of Integer, CGroup)
    Private ReadOnly Property Index() As Dictionary(Of Integer, CGroup)
        Get
            If Not IsNothing(_index) Then
                If _index.Count = Me.Count Then
                    Return _Index
                End If
            End If
            _index = New Dictionary(Of Integer, CGroup)(Me.Count)
            For Each i As CGroup in Me
                _index(i.GroupId) = i
            Next
            Return _index
        End Get
    End Property
#End Region
    
#Region "Foreign-Key Indices (Subsets)"
#End Region

#Region "Move Up/Down"
    Public Sub MoveUp(ByVal s As CGroup)
        Move(s, -1)
    End Sub
    Public Sub MoveDown(ByVal s As CGroup)
        Move(s, 1)
    End Sub

    'Private
    Private Sub Move(ByVal s As CGroup, ByVal change As Integer)
        If IsNothing(s) Then Exit Sub
        Dim index As Integer = Me.IndexOf(s) + change
        If index < 0 Then Exit Sub
        If index > Me.Count - 1 Then Exit Sub

        'Modify a copy of the array for threadsafety
        Dim dd As New CGroupList(Me.Count)
        dd.AddRange(Me)
        With dd
            .Remove(s)
            .Insert(index, s)
            .ResetOrdinals()
        End With

        Me.SaveAll()
    End Sub
    Private Sub ResetOrdinals()
        For i As Integer = 0 To Me.Count - 1
            Me(i).GroupSortOrder = i
        Next
    End Sub
#End Region

End Class